File: frontend/lib/renderedPercent.ts
import { leaderFirstSlice } from "./discover/leaderOrder";

// The whole percent this surface prints for a probability — web's arm of
// `contracts/rendered_percent.json` (#1933).
//
// This was `Math.round(probability * 100)` inline in the Label Pass page, which
// is correct and was never the bug. It is extracted because it is now half of a
// CROSS-RUNTIME contract: the server takes the graded card's fingerprint at this
// exact resolution so that a drift refusal is always explicable to the person
// who was looking at the card, and Swift prints the same number on native. An
// expression inlined in a JSX attribute cannot be driven through a shared table;
// a named function can, and all three arms now are.
//
// UX-P110 shipped the Python side using banker's rounding against this, and the
// test beside it asserted the JavaScript answer in a comment while expecting the
// Python one in the assertion. See the contract file for why a comment was never
// going to hold this together.

export function renderedPercent(probability: number | null | undefined): number | null {
  if (probability === null || probability === undefined) return null;
  if (!Number.isFinite(probability)) return null;
  return Math.round(probability * 100);
}

// ── The card-level half of the contract (#2060) ──────────────────────────────
//
// Getting each number right is not enough, because a surface prints a CARD and a
// card has a SUM. `Los Angeles D 0.925 / Colorado 0.075` renders 93 and 8 above —
// both correct, 101 together — because Kalshi quotes a complement pair on a
// half-cent grid, so `p * 100` lands on `.5` for both sides at once and half-up
// rounds both up.
//
// See `contracts/rendered_percent.json` for the measured population behind the
// [0.99, 1.01] band and for why `field_coherence`'s much looser [0.5, 1.5] is
// deliberately not the predicate.

const COMPLEMENT_MIN = 0.99;
const COMPLEMENT_MAX = 1.01;

export function isComplementPair(
  probabilities: Array<number | null | undefined> | null | undefined,
): boolean {
  if (!probabilities || probabilities.length !== 2) return false;
  const values = probabilities.filter(
    (p): p is number => p !== null && p !== undefined && Number.isFinite(p),
  );
  if (values.length !== 2) return false;
  const total = values[0] + values[1];
  return total >= COMPLEMENT_MIN && total <= COMPLEMENT_MAX;
}

/**
 * The whole percents this surface prints for ONE CARD's served outcomes.
 *
 * A complement pair is normalized by its true total (removing the vig
 * symmetrically rather than dumping it all on one side), index 0 is rounded once
 * with `renderedPercent`, and index 1 is DERIVED as `100 - index0`. Index 0 is the
 * card's headline, so it is the value that survives untouched.
 *
 * Everything else is rendered exactly as before — that direction is asserted by
 * the contract suite as explicitly as the fixed one.
 */
export function renderedCardPercents(
  probabilities: Array<number | null | undefined> | null | undefined,
): Array<number | null> {
  if (!probabilities || probabilities.length === 0) return [];
  if (!isComplementPair(probabilities)) return probabilities.map(renderedPercent);

  const total = (probabilities[0] as number) + (probabilities[1] as number);
  const leader = renderedPercent((probabilities[0] as number) / total);
  if (leader === null) return probabilities.map(renderedPercent);
  return [leader, 100 - leader];
}

// ── Why a card's two numbers do not add up (#2088) ───────────────────────────
//
// `renderedCardPercents` fixed the pair that SHOULD total 100 and left alone the
// pair that should not — correctly, because normalizing a pair summing to 0.97
// would invent three points of probability. But it left the reader a card reading
// `57 / 40` with nothing saying why, which looks exactly like the `93 / 8` bug it
// had just fixed. #2088: an unexplained non-100 is the defect, a labelled one is a
// fact.
//
// THE SERVER DECIDES THIS (`card_sum_reason` on both labeling serializers). These
// are the FALLBACK for a payload minted before that field existed, and they are in
// the contract so the fallback cannot drift from the served answer — the same role
// `renderedPercent` plays for a pre-#2060 payload.
//
// See `contracts/rendered_percent.json` (`card_sum_cases`) for the measured
// population, for why the reason is NOT an illiquidity mark (the exemplar's book
// grades `traded`), and for why arity other than two is null.

/** A served outcome has no price at all, so there is no total to check. */
export const SUM_UNPRICED_OUTCOME = "unpriced_outcome";
/** Both sides priced, outside the complement band — two independent questions. */
export const SUM_INDEPENDENT_PRICES = "independent_prices";

/**
 * The integer total this surface prints for one card, or null if it prints nothing.
 *
 * An unpriced outcome contributes nothing rather than a zero — "no price" and "0%"
 * are different cards — so `[57, null]` totals 57 and is explained by
 * `cardSumReason` rather than reported as a 43-point miss.
 */
export function cardSum(
  probabilities: Array<number | null | undefined> | null | undefined,
): number | null {
  const percents = renderedCardPercents(probabilities).filter(
    (p): p is number => p !== null,
  );
  if (percents.length === 0) return null;
  return percents.reduce((a, b) => a + b, 0);
}

/**
 * Why this card's printed percents do not total 100, or null if they do.
 *
 * Taken over `renderedCardPercents` rather than the raw floats, so it answers for
 * THE PICTURE. A complement pair is normalized, rounded once and derived, so it
 * totals 100 by construction and can never earn a reason.
 *
 * null for any arity other than two: it means "no claim about a total is made
 * here", never "checked and fine", and a surface must not render it as one.
 */
export function cardSumReason(
  probabilities: Array<number | null | undefined> | null | undefined,
): string | null {
  if (!probabilities || probabilities.length !== 2) return null;
  const percents = renderedCardPercents(probabilities);
  if (percents.some((p) => p === null)) return SUM_UNPRICED_OUTCOME;
  const total = percents.reduce((a, b) => (a ?? 0) + (b ?? 0), 0);
  return total === 100 ? null : SUM_INDEPENDENT_PRICES;
}

// ── The duel: the same question in FIXED positions (UX-P114) ─────────────────
//
// `renderedCardPercents` assumes served order, where index 0 is the headline. The
// Discover event card does not sort — away is always left, home is always right —
// yet it is the most exact complement pair in the product, because the feed derives
// the away side as `1 - home`. So whenever `home * 100` lands on `.5`, both sides
// round up and the strip prints 101. Measured 2026-08-21: 34 of 414 live/upcoming
// events (8.2%), all 101, never 99.
//
// THE SERVER NOW DECIDES THIS (`current_odds.{away,home}_rendered_percent`), because
// four surfaces draw this strip and only one of them is this file. This function is
// the local FALLBACK for a payload from before that field existed — and it is in the
// contract so the fallback cannot drift from the served answer.
//
// See `contracts/rendered_percent.json` (`duel_cases`) for why the FAVOURITE is the
// side that survives rather than the away side.

export function renderedDuelPercents(
  awayProbability: number | null | undefined,
  homeProbability: number | null | undefined,
): Array<number | null> {
  const pair = [awayProbability, homeProbability];
  if (!isComplementPair(pair)) {
    return [renderedPercent(awayProbability), renderedPercent(homeProbability)];
  }

  const away = awayProbability as number;
  const home = homeProbability as number;
  if (away >= home) return renderedCardPercents([away, home]);
  const [homePct, awayPct] = renderedCardPercents([home, away]);
  return [awayPct, homePct];
}

// ── The HEADLINE percent, so two surfaces cannot headline one market twice ───
//
// UX-P162. `renderedCardPercents` answers for a LIST; a card's hero prints ONE
// number out of that list, and picking it correctly is three separate decisions
// that were spelled out longhand in `FeedCard` and not at all in Discover:
//
//   1. the served percent wins, keyed on the KEY BEING ABSENT rather than on the
//      value being falsy — `?? derive()` re-derives on every correct card and
//      makes the server's answer decorative;
//   2. the rule is anchored on the LEADER-FIRST slice, because
//      `renderedCardPercents` leaves index 0 untouched and derives index 1, so
//      whichever row is first is the one that survives rounding. Anchoring on
//      served order would let two surfaces normalize a pair around opposite ends
//      and disagree by a point — the exact failure this function prevents;
//   3. the hero is then found BY IDENTITY, not by position, because the headline
//      is `top_outcomes[0]` (served order) while the slice is leader-first, so
//      the two indices are not the same list.
//
// Null means "no override" — the caller keeps `formatProbabilityPercent`'s own
// rounding — and that is also the answer when the headline is not among the
// printed rows at all, rather than inventing a number for it.
//
// MEASURED on the deployed feed 2026-08-29 (114 unique futures cards across
// `/api/feed` plus the politics, economics, entertainment and sports surfaces):
// 7 are two-outcome and 0 disagree today, so this is a LATENT fix, not a visible
// one. It is still real — a pair summing to 1.005 shifts the leader by a point
// under the rule and not under raw rounding — and 1 of the 103 multi-outcome
// cards already ships `top_outcomes[0]` that is NOT the maximum, which is what
// makes decision 3 load-bearing rather than defensive.

/** The shape both feed serializers give an outcome the card can print. */
type RenderableOutcome = {
  probability?: number | null;
  rendered_percent?: number | null;
};

export function renderedLeaderPercent(
  topOutcomes: ReadonlyArray<RenderableOutcome> | null | undefined,
  leader: RenderableOutcome | null | undefined,
  printedCount = 3,
): number | null {
  if (!leader) return null;
  if ("rendered_percent" in leader) return leader.rendered_percent ?? null;

  const printed = leaderFirstSlice(topOutcomes ?? [], printedCount);
  const index = printed.indexOf(leader);
  if (index < 0) return null;
  return renderedCardPercents(printed.map((o) => o.probability))[index] ?? null;
}

File: ios/Bain Luck/Bain Luck/Utilities/RenderedPercent.swift
import Foundation

/// The whole percent this app prints for a probability — native's arm of
/// `contracts/rendered_percent.json` (#1933).
///
/// ## Why a one-line function has a contract behind it
///
/// The server takes a graded card's drift fingerprint at exactly this
/// resolution, so that a refused judgment is always explicable to the person who
/// was looking at the card: "the number on screen changed". If native prints 57
/// where the server fingerprinted 56, that sentence is false at precisely the
/// values where it is hardest to be right — the verdict is refused for drift
/// nobody can see, or accepted against a card that visibly moved.
///
/// Three runtimes print this number and no import spans them, so the shared unit
/// is the table in `contracts/rendered_percent.json` and each implementation is
/// driven through every row of it (ruling 021). `RenderedPercentContractTests`
/// executes this arm; the jest contract suite asserts that test's case table
/// still equals the contract, because the Swift gate does not run in CI and a
/// runtime check nobody runs is not a check.
///
/// `Double.rounded()` defaults to `.toNearestOrAwayFromZero`, which is half-up
/// over the non-negative domain probabilities live in — the same rule as
/// JavaScript's `Math.round` and as the server's `floor(x + 0.5)`, and
/// deliberately NOT Python's built-in `round`, which is banker's rounding and
/// shipped wrong once already.
///
/// The scaling happens BEFORE the rounding, in `Double`. That is load-bearing.
///
/// ## Why the scale is 1000/10 and not 100 (#3867, contract version 5)
///
/// This comment used to say the opposite: that `0.565 * 100` being
/// `56.49999999999999` made **56** "the honest answer in all three runtimes", and
/// that anything correcting it "with a decimal type would print 57 and silently
/// leave the contract". The parity worry was right and is preserved — the scaling
/// worry was not.
///
/// `* 100` is half-up on the double the wire value BECAME, not on the number it
/// SENT. The venues quote on a half-percent grid, so `.xx5` is the common case
/// here, and `* 100` lands a hair below the boundary for some of those values and
/// exactly on it for others: `0.585` printed 59 while `0.565` printed 56 — one
/// half rounding up, its neighbour down, for a reason no reader can see. Scaling
/// by 1000 first recovers the quoted decimal, so a half rounds up, always.
///
/// This is still `Double`, still the same operations as the other two runtimes,
/// so it is NOT the decimal type the old note warned about — parity was
/// re-measured, not assumed. Exactly four three-decimal wire values move: 0.145,
/// 0.285, 0.565, 0.575. And the rule changed in the CONTRACT first, which is the
/// only way this line is allowed to change.
nonisolated func renderedPercent(_ probability: Double?) -> Int? {
    guard let probability, probability.isFinite else { return nil }
    return Int((probability * 1000 / 10).rounded())
}

/// A two-outcome field is a complement pair when its members sum into this band.
///
/// 1.01 is not a new number: `card_integrity.display_scale` already uses exactly
/// it as the two-outcome "true binary" threshold. This is that constant made
/// SYMMETRIC, and the missing lower half was itself half the defect — a pair
/// summing to 0.99 printed 99 and nothing in the system considered that wrong.
private let complementMin = 0.99
private let complementMax = 1.01

nonisolated func isComplementPair(_ probabilities: [Double?]) -> Bool {
    guard probabilities.count == 2 else { return false }
    let values = probabilities.compactMap { $0 }.filter { $0.isFinite }
    guard values.count == 2 else { return false }
    let total = values[0] + values[1]
    return total >= complementMin && total <= complementMax
}

/// The whole percents this app prints for ONE CARD's served outcomes — the
/// card-level half of `contracts/rendered_percent.json` (#2060).
///
/// ## Why the scalar rule above was not enough
///
/// A surface prints a CARD, and a card has a SUM. Alex's 08-20 gold session served
/// `Los Angeles D 0.925 / Colorado 0.075`, which `renderedPercent` turns into 93
/// and 8 — both correct, and 101 together. Kalshi quotes a complement pair on a
/// HALF-CENT grid, so `p * 100` lands on `.5` for **both sides at once** and
/// half-up rounds both up. Measured on production 2026-08-21, 10,198 of 21,524
/// open two-outcome markets rendered a sum other than 100, 8,982 of them at 101.
///
/// A complement pair is normalized by its true total, index 0 is rounded ONCE, and
/// index 1 is DERIVED as `100 - index0`. Index 0 is the card's headline, so it is
/// the number that survives untouched and the derived point lands on the side
/// nobody is quoting. Everything that is not a complement pair renders exactly as
/// before — the contract table pins that direction as hard as the other.
nonisolated func renderedCardPercents(_ probabilities: [Double?]) -> [Int?] {
    guard !probabilities.isEmpty else { return [] }
    guard isComplementPair(probabilities),
          let first = probabilities[0], let second = probabilities[1] else {
        return probabilities.map { renderedPercent($0) }
    }
    let total = first + second
    guard let leader = renderedPercent(first / total) else {
        return probabilities.map { renderedPercent($0) }
    }
    return [leader, 100 - leader]
}

/// The two whole percents a GAME card prints, returned as `[away, home]` —
/// version 3 of `contracts/rendered_percent.json` (UX-P114).
///
/// ## Why the card rule above needed a positional sibling
///
/// `renderedCardPercents` assumes SERVED ORDER, where index 0 is the headline
/// because the labeling serializers sort descending first. A game card does not
/// sort: away is always drawn left and home always right, because those positions
/// carry meaning a probability ranking would destroy.
///
/// It is still the most exact complement pair in the product — `routes/feed.py`
/// derives the away side as `round(1.0 - current_home_prob, 6)` — so the defect
/// fires on a provable condition: when `home * 100` lands exactly on `.5`, both
/// sides round up and the strip prints 101. It can never print 99. Measured on
/// production 2026-08-21 over the 414 scheduled/live events in the feed's window:
/// 34 (8.2%) printed 101, including Green Bay @ Denver and Toronto FC @ Inter
/// Miami.
///
/// ## The favourite is the side that survives
///
/// A duel has no served order to inherit a headline from, so the replacement rule
/// is the one the card rule expresses: the number a reader anchors on is left
/// untouched and the derived point lands on the underdog. Always-away-first would
/// instead move the favourite half the time — on Green Bay @ Denver it prints 67
/// for a side whose own correct value is 68.
///
/// The SERVER now decides this and sends it as
/// `current_odds.{away,home}_rendered_percent`, because four surfaces draw this
/// strip. This function is the local fallback for a payload from before that field
/// existed, and it lives in the contract so the fallback cannot drift from the
/// served answer.
nonisolated func renderedDuelPercents(
    away awayProbability: Double?,
    home homeProbability: Double?
) -> [Int?] {
    let pair = [awayProbability, homeProbability]
    guard isComplementPair(pair),
          let away = awayProbability, let home = homeProbability else {
        return [renderedPercent(awayProbability), renderedPercent(homeProbability)]
    }
    if away >= home {
        return renderedCardPercents([away, home])
    }
    let flipped = renderedCardPercents([home, away])
    return [flipped[1], flipped[0]]
}

/// The two whole percents a game strip prints, choosing between the SERVED pair
/// and the local one — `[away, home]`, and the last word on both (#2279).
///
/// ## Both served values or neither
///
/// UX-P114 gave every game-card surface `current_odds.{away,home}_rendered_percent`
/// so the two sides of one question are decided ONCE, by the server. Three surfaces
/// adopted it and all three coalesced **per side**:
///
///     let awayPct = odds.awayRenderedPercent ?? duelFallback[0]
///     let homePct = odds.homeRenderedPercent ?? duelFallback[1]
///
/// A payload carrying one field and not the other therefore prints a served value
/// beside a locally derived one, and that re-opens the very 101 UX-P114 shipped to
/// close — from the other direction. On `0.505 / 0.495` the served home is 51 and a
/// naively derived away is 50. The struct's own comment
/// (`Models/CommonTypes.swift`) says these fields are optional precisely because a
/// Discover response is CACHED and this build can be installed against an older
/// deploy; a response written across a partial rollout is the case the fallback
/// exists for, and it is exactly the case the per-side form gets wrong.
///
/// So the two served values are ONE decision. Either both are present and both are
/// used, or the pair falls back WHOLE to `renderedDuelPercents`. The web arm states
/// the same rule in `lib/eventKeyStats.ts` and the event page states it inline;
/// this is the native surfaces' shared copy of it (ruling 021 — share the DECISION,
/// not the ingredient).
///
/// ## The served pair describes `current_odds` AND NOTHING ELSE
///
/// A caller whose probabilities came from somewhere else — `opening_odds`, a
/// history row, a chart point — must pass `nil` for both served values. Handing in
/// `current_odds`' rounding beside another source's probability prints a mismatched
/// pair that still sums to 100, so no sum guard can see it. `servedAway`/`servedHome`
/// are separate parameters rather than a `CurrentOdds` so that the caller has to
/// make that choice at the branch that knows the answer.
nonisolated func duelPercents(
    away awayProbability: Double?,
    home homeProbability: Double?,
    servedAway: Int?,
    servedHome: Int?
) -> [Int?] {
    if let servedAway, let servedHome {
        return [servedAway, servedHome]
    }
    return renderedDuelPercents(away: awayProbability, home: homeProbability)
}

/// The one whole percent a compact row prints beside an away-first
/// `"<away> vs <home>"` title — the FIRST-NAMED side's.
///
/// ## Why a lone number needs a rule at all
///
/// #4306: `SearchView.searchEventRow` built its title away-first — the house
/// convention, 6 of the 7 sites that build one — and then drew
/// `homeProbability` beside it with no label. A reader anchors a lone number to
/// the name they read first, so every scheduled row stated the opposite of the
/// market. Measured on production 2026-09-09, 6 of 6 US Open rows, worst case
/// "Botic van de Zandschulp vs Alexander Zverev — 86%" for a side priced at
/// 13.8%. The finished arm of that same function already renders its score
/// away-first, so one row disagreed with itself depending on whether the match
/// had started.
///
/// ## Why it returns a percent and not a probability
///
/// The number a reader compares against the event page they tap through to has
/// to be the same INTEGER that page drew, and `duelPercents` is the last word on
/// that (#2279). Rounding again here would reintroduce the off-by-one from the
/// other direction.
///
/// ## A payload with only the home side still gets a number
///
/// `away_probability` accompanied `home_probability` on 53 of 53 rows carrying
/// odds across four production search queries (2026-09-09), so this is a
/// belt-and-braces path rather than a live case. It exists because the row it
/// replaces drew a number whenever HOME was present, and a fix for a
/// wrong-sided number must not turn into a missing one: `renderedDuelPercents`
/// answers `nil` for a pair that is not a complement, which is what a
/// home-only payload is. Deriving the complement is the move the two other
/// single-sided readers already make (`MenuBarView`, `TeamDetailView`).
///
/// The derivation is here and not at the call site so that the rule and its
/// edge case are one testable thing — CI compiles no Swift (#4302), so a
/// branch that lives in a view body is a branch no gate can reach.
/// Returns the probability as well as the percent because the caller needs
/// both and must not derive either a second time: `formatProbability` runs its
/// `<1%` / `>99%` guards on the PROBABILITY (they are a claim about the value,
/// not about the rounding), so a call site holding only the integer cannot
/// format the number correctly, and a call site deriving the Double itself is a
/// second copy of the rule this function exists to hold.
nonisolated func firstNamedSideNumber(
    away awayProbability: Double?,
    home homeProbability: Double?,
    servedAway: Int?,
    servedHome: Int?
) -> (probability: Double, percent: Int)? {
    guard let away = awayProbability ?? homeProbability.map({ 1 - $0 }) else { return nil }
    guard let percent = duelPercents(
        away: away,
        home: homeProbability,
        servedAway: servedAway,
        servedHome: servedHome
    )[0] else { return nil }
    return (probability: away, percent: percent)
}

/// The same lone number, for a row that may be priced by `current_odds` OR by the
/// blend the server already computed — preferring `current_odds` when both exist.
///
/// ## Why the fallback needs a function and not an `??` at the call site
///
/// #4967: `SearchView.searchEventRow` drew a percentage only when `current_odds`
/// was present, so an unfinished row whose price the server HAD already blended
/// printed nothing at all. Measured on production 2026-09-10 across eight
/// queries: of 84 unfinished rows, 29 drew no number, and 20 of those 29 carried
/// a complete `hero_probability` / `hero_probability_away` pair the phone was
/// throwing away. The remaining 9 carry neither and must keep printing nothing
/// (#4794 — a row with no reading states no reading).
///
/// ## The served percents belong to `current_odds`, so the hero arm cannot see them
///
/// `duelPercents`' contract above is explicit that `servedAway`/`servedHome`
/// describe `current_odds` and nothing else: pairing that rounding with another
/// source's probability prints a mismatched pair that still sums to 100, which no
/// sum guard can catch. The hero arm therefore passes `nil` for both and lets
/// `renderedDuelPercents` derive. That is enforced HERE rather than trusted at the
/// call site, because the call site is a view body and CI compiles no Swift
/// (#4302) — a rule that lives in a `View` is a rule no gate can reach.
///
/// Preference order is `current_odds` first so that no row which renders correctly
/// today changes: the hero pair is strictly a fallback for rows drawing nothing.
nonisolated func firstNamedSideNumber(
    oddsAway: Double?,
    oddsHome: Double?,
    servedAway: Int?,
    servedHome: Int?,
    heroAway: Double?,
    heroHome: Double?
) -> (probability: Double, percent: Int)? {
    if let fromOdds = firstNamedSideNumber(
        away: oddsAway,
        home: oddsHome,
        servedAway: servedAway,
        servedHome: servedHome
    ) {
        return fromOdds
    }
    return firstNamedSideNumber(
        away: heroAway,
        home: heroHome,
        servedAway: nil,
        servedHome: nil
    )
}
